### DISCUSSION TITLE
CharacterVirtual in moving and rotating object

### DISCUSSION THREAD

**Discussion by DocAce** [jrouwe/JoltPhysics#395] [Q&A] [answered]
In the game I'm working on you can walk inside spaceships while they're flying around. It should feel the same as if the ship was completely stationary. I've implemented player movement using CharacterVirtual but I'm having some problems getting it to work correctly.

The CharacterVirtual as well as the ship's Body are in the same PhysicsSystem as all other objects, so the gravity vector can change if the ship is moving.

I'm focusing on the case of a stationary but rotated ship at the moment. I'm calling SetUp() each frame, before calling the update function, with the correct up vector of the ship relative to the physics coordinate system. The gravity vector I'm passing in is also rotated to match the ship's down vector in the physics coordinate system. I'm also making sure that the vectors in ExtendedUpdateSettings (like mStickToFloorStepDown) are rotated accordingly. Should this be enough to support this kind of setup?

The problems I'm having are largely related to jumping and gravity, both of which appear to lose strength as the up vector gets less aligned with the global up vector.

Edit: I have now also tested the case of a ship that doesn't rotate but moves quite slowly (10m/s) and without acceleration. I can walk around for the most part, but I again get stuck where I wouldn't on a stationary ship, and I also sometimes get odd vertical position jumps. I already ruled out StickToFloor and WalkStairs by setting the vectors to zero. Any ideas?

**Comment by jrouwe** [comment]
Hello,

I tried to reproduce the issue, but as far as I can see everything works as intended. I've created #396 to be able to set the up vector in the CharacterVirtualTest. Can you repro your issue in that test?

I haven't tested a fast moving 'ship' which is a far more tricky case. Because the simulation runs at a different time than the character itself, the character only sees an updated ship position, so if the ship moves fast enough the character can easily teleport halfway or entirely through a wall (tunnelling). If you get far enough into a wall it's difficult to determine what's inside and what's outside so you can indeed get stuck. I think you should probably add the delta position/rotation of the ship to the character before calling the update function so that the character movement becomes relative to the ship. If you need the acceleration of the ship to affect the character, I would express that by adding an extra velocity to the character.

**Comment by jrouwe** [comment]
B.t.w. rotating the character can also cause issues: If you do this without checking for collision, it is again possible that you rotate the character into geometry and if the geometry is complex enough and the intersection deep enough then it is possible that you get stuck (if the generated contacts have conflicting normals).

**Comment by DocAce** [comment]
The character's movement happens completely relative to the moving ship. The ship's acceleration shouldn't affect the character, you can assume 100% perfect inertial compensators :)
On our side of the code the character position is stored in coordinates relative to the ship. Before calling ExtendedUpdate() the VirtualCharacter's position and rotation are set to the character's offset plus the ship's offset. This happens right after the physics simulation so all body offsets are up-to-date.

Here is the (simplified) code:
```
PosRot physicsoffset(characteroffset.Transform(shipoffset));
JoltCharacter->SetPosition(physicsoffset.GetPos());
JoltCharacter->SetRotation(physicsoffset.GetRot());

JoltCharacter->SetUp(shipoffset.GetRot().Up());

Vector3 gravity = Vector3(0.0f, -Gravity * dt, 0.0f);

Vector3 velocity(Zero);
if (OnGround) {
	if (Jump) {
		velocity += Vector3(0.0f, JumpSpeed, 0.0f);
	}
}
else {
	velocity = Vector3(0.0f, JoltCharacter->GetLinearVelocity().GetY(), 0.0f);
	inputdirection.SetY(0.0f);
}
velocity += gravity;
velocity += inputdirection;
JoltCharacter->SetLinearVelocity(velocity.Rotate(shipoffset.GetRot()));

JPH::CharacterVirtual::ExtendedUpdateSettings update_settings;
update_settings.mStickToFloorStepDown = JPH::Vec3::sZero();
update_settings.mWalkStairsStepUp = JPH::Vec3::sZero();
update_settings.mWalkStairsMinStepForward = 0.02f;
update_settings.mWalkStairsStepForwardTest = 0.15f;
update_settings.mWalkStairsCosAngleForwardContact = Math::Cos(DEG2RAD(75.0f));
update_settings.mWalkStairsStepDownExtra = JPH::Vec3::sZero();

JoltCharacter->ExtendedUpdate(dt, gravity.Rotate(shipoffset.GetRot()), update_settings, ...);
```

I've recorded a few short clips to show the actual problems I'm having.

The working case, in a non-moving ship that is aligned with the coordinate space of the physics system:
https://youtu.be/7dwVoho5Fek
As you can see, walking and jumping is perfectly smooth.

Now with the ship rotated by around 45 degrees around the z axis:
https://youtu.be/Da2M199NBqw
There's a really strange bug when I walk over the downward slope. (I first suspected StickToFloor(), but mStickToFloorStepDown was zero when I recorded these.) You can also see that the jump height is reduced.

Here's the ship rotated by 180 degrees:
https://youtu.be/2vQvj2po1RU
You can clearly see where I'm trying to jump (I basically go back to the floor in the next frame and start jumping again), and the slowed down gravity is also at its most extreme.

And with a moving ship:
https://youtu.be/eKAjfxjKmt0
I can walk around fine in most places, but get stuck at slopes, it seems like.

**Comment by jrouwe** [comment, marked as answer]
Hello,

I have the feeling that there's an issue with the velocity calculation:

```
velocity = Vector3(0.0f, JoltCharacter->GetLinearVelocity().GetY(), 0.0f);
```

GetLinearVelocity().GetY() will not return the vertical velocity when the ship is rotated, this will:

```
velocity = Vector3(0.0f, JoltCharacter->GetLinearVelocity().Dot(JoltCharacter->GetUp()), 0.0f);
```

Also I don't see anything to fix up the character velocity when the ship has rotated during its update, so a better solution is probably:

```
velocity = Vector3(0.0f, JoltCharacter->GetLinearVelocity().Dot(previous_frame_up_vector), 0.0f);
```

Where previous_frame_up_vector can simply be gotten before you call JoltCharacter->SetUp(...).

I'm assuming characterOffset is calculated after calling ExtendedUpdate by transforming the character position back relative to the ship and then storing it until the next frame?

**Reply by DocAce** [reply]
Oh my god, what an embarrassing mistake, thank you!
Anyone with impostor syndrome, take note: I have been doing this job professionally for over 15 years... 😆

> I'm assuming characterOffset is calculated after calling ExtendedUpdate by transforming the character position back relative to the ship and then storing it until the next frame?

Not quite exactly, but in principle yes.

**Comment by DocAce** [comment]
To follow up on the velocity issue: it happens even if the ship is moving very slowly (1m/s). I end up being unable to walk up even a minor slope if it's going in the direction of ship movement. I've tracked this to CharacterVirtual::HandleContact(), which subtracts the contact's linear velocity from InVelocity. For the time being I can manage this by simply ignoring that velocity, but I wonder if some way of setting a kind of reference velocity to be added to relative_velocity, cancelling out the contact's velocity, could be useful there?

**Reply by jrouwe** [reply]
Yes, you're right. I will try to create a demo of this use case myself. I think a simple solution could be to pass in the contact velocity to CharacterContactListener::OnContactAdded and to allow the application to override it (i.e. subtract the reference frame velocity). But I need to play around a bit to see if this is the best approach.

**Reply by jrouwe** [reply]
I created a way to override the speed of a body and a demo:

https://github.com/jrouwe/JoltPhysics/tree/feature/inertial_damping

I'm not sure if I'm 100% happy with it yet, but please try it out.

**Reply by jrouwe** [reply]
Merged the change in #408
